peer.go

```diff
@@ -0,0 +1,1087 @@
+				p.shared.options.failureWait,
+				p.shared.options.maxFailureWait))
+		}
+		rpc.orig.process(p, rpc.err)
+
+	default:
+		p.sendProgress = true
+		p.failures = 0
+		p.backoffTimer.Stop()
+		rpc.process(p, nil)
+		p.progress.lastContact = rpc.started()
+		p.progress.lastReply = time.Now()
+	}
+}
+
+// issueWork does a single unit of non-blocking work.
+func (p *peerState) issueWork() (didSomething bool) {
+	if p.progress.term > p.control.term {
+		return false
+	}
+
+	switch p.control.role {
+	case Shutdown:
+		// do nothing
+
+	case Follower:
+		// do nothing
+
+	case Candidate:
+		// Send a RequestVote RPC.
+		if !p.sentRequestVote && p.failures == 0 {
+			p.start(makeRequestVoteRPC)
+			return true
+		}
+
+	case Leader:
+		// Send a heartbeat (empty AppendEntries) RPC.
+		if time.Now().After(p.lastHeartbeatSent.Add(p.shared.options.heartbeatInterval)) {
+			p.start(makeHeartbeatRPC)
+			return true
+		}
+
+		// Send an AppendEntries/InstallSnapshot RPC.
+		if p.nextIndex < p.control.lastIndex+1 && p.failures == 0 {
+			if p.needsSnapshot {
+				if !p.outstandingInstallSnapshotRPC {
+					p.start(makeInstallSnapshotRPC)
+					return true
+				}
+			} else {
+				if p.outstandingAppendEntriesRPCs == 0 ||
+					(p.allowPipeline && !p.outstandingPipelineSend) {
+					p.start(makeAppendEntriesRPC)
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func (p *peerState) enablePipeline() {
+	if !p.allowPipeline {
+		p.allowPipeline = true
+		p.shared.logger.Printf("[INFO] raft: Enabling pipelining replication to peer %v",
+			p.shared.peerID)
+	}
+}
+
+func (p *peerState) disablePipeline() {
+	if p.allowPipeline {
+		p.allowPipeline = false
+		p.shared.logger.Printf("[INFO] raft: Disabling pipeline replication to peer %v",
+			p.shared.peerID)
+	}
+}
+
+// updateTerm clears out some progress fields when its term changes.
+func updateTerm(progress *peerProgress, term uint64) {
+	if term > progress.term {
+		progress.term = term
+		progress.voteGranted = false
+		progress.matchIndex = 0
+	}
+}
+
+// Every type of RPC sent to peers implements peerRPC. Some of its methods are
+// invoked on the main Peer goroutine and may not block. Others are run on
+// separate routines and may block. A typical RPC goes through the following
+// lifecycle:
+//
+//       selectLoop            helper routine
+//      (nonblocking)            (blocking)
+//      =============          ==============
+//
+//        make
+//          |
+//          '-------spawn-----> prepare
+//                                 |
+//       confirm <---requestCh-----'
+//          |
+//         (ok)
+//          |
+//          '-------spawn-----> sendRecv
+//                                 |
+//                                 '------------> Transport
+//                                                    |
+//                                 .------------------'
+//                                 |
+//       process <--replyCh--------'
+//
+type peerRPC interface {
```

Comment:
If I get what you're saying, it didn't seem advantageous to me when I tried. Feel free to give it a shot (now or later).